include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,537 @@
+     */
+    /*!\brief Compare seqan3::gap_decorator_anchor_set by underlying sequence and gaps.
+     * \tparam    inner_type2  The underlying type of the decorator to compare with; must be the same as inner_type
+     *                         when ignoring constness.
+     * \param[in] rhs          The right-hand side gap decorator to compare.
+     * \returns A boolean flag indicating (in)equality of the aligned sequences.
+     *
+     * \par Complexity
+     * Worst case: \f$O(n*\log k)\f$
+     *
+     * \par Exceptions
+     *
+     * No-throw guarantee. Does not modify the aligned sequences.
+     */
+    template <typename inner_type2>
+    //!\cond
+        requires std::is_same_v<std::remove_const_t<inner_type>, std::remove_const_t<inner_type2>>
+    //!\endcond
+    constexpr bool operator==(gap_decorator_anchor_set<inner_type2> const & rhs) const noexcept
+    {
+        if (anchors.size() != rhs.anchors.size())
+            return false;
+
+        if (sequence) // not nullptr
+        {
+            if (rhs.sequence) // not nullptr
+            {
+                if (sequence->size() == rhs.sequence->size())
+                    return std::ranges::equal(*this, rhs);
+            }
+        }
+        else if (!rhs.sequence)
+        {
+            return true; // both are nullptr (and both anchors should be empty as inserting gaps into a nullptr is UB)
+        }
+
+        return false;
+    }
+
+    //!\copydoc operator==
+    template <typename inner_type2>
+    //!\cond
+        requires std::is_same_v<std::remove_const_t<inner_type>, std::remove_const_t<inner_type2>>
+    //!\endcond
+    constexpr bool operator!=(gap_decorator_anchor_set<inner_type2> const & rhs) const noexcept
+    {
+        return !(operator==(rhs));
+    }
+    //!\}
+private:
+    //!\brief The gap type as a tuple storing position and accumulated gap lengths.
+    using gap_t = typename std::pair<size_t, size_t>;
+
+    /*!\brief Structure allowing the comparison of gaps.
+     * \tparam        gap_t Type of the gap storing position an length accumulator.
+     *
+     * \details
+     *
+     * It is assumed that the gap structure is always in a consistent
+     * state, i.e. there are no two gaps that are overlapping. The ordering of gaps
+     * structures is exclusively dependent on the gap starting position which is
+     * the ordering criterion for the anchor set implemented as ordered red-black tree.
+     */
+    template <typename gap_t>
+    struct gap_compare {
+        /*!\brief The operator allowing gap_t comparison.
+         * \param[in] lhs   The left-hand gap to compare.
+         * \param[in] rhs   The right-hand side gap to be compared.
+         * \returns A boolean flag indicating whether the gap starting position
+         *          of \p lhs is smaller than the one of \p rhs.
+         */
+         bool operator() (const gap_t& lhs, const gap_t& rhs) const {
+             return lhs.first < rhs.first;
+         }
+     };
```

Comment:
Why do we have this? Why not just a lambda function?

Reply:
If we need this, correct placement of braces

Reply:
I removed it in favour of a `dummy`_bound` that can be used when searching with lower_bound() because we need to make sure that the second pair value is never smaller. This is the same as overloading the comparator.